File: Codeforces/580C.cpp
#include <bits/stdc++.h>
using namespace std;

vector<int> adj[1000500];
vector<char> used(1000500);
int cat[1000500], n, m;
int ans = 0;

void dfs(int v, int cnt=0) {
    cnt = cat[v] == 1 ? cnt + 1 : 0;
    if (cnt > m) {
        return;
    }
    used[v] = true;
    if ((int)adj[v].size() == 1 && v != 1) {
        ans++;
    }
    for (auto i : adj[v]) {
        if (!used[i]) {
            dfs(i, cnt);
        }
    }
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    cin >> n >> m;
    for (int i = 1; i <= n; i++) {
        cin >> cat[i];
    }
    for (int i = 1; i < n; i++) {
        int u, v;
        cin >> u >> v;
        adj[u].push_back(v);
        adj[v].push_back(u);
    }

    dfs(1);
    cout << ans << "\n";
}

File: Codeforces/580C.py
n, m = map(int, input().split())
cat = list(map(int, input().split()))
cat = [0] + cat
adj = [[] for _ in range(n + 1)]

for _ in range(n - 1):
    u, v = map(int, input().split())
    adj[u].append(v)
    adj[v].append(u)

visited = set()
ans = 0

def dfs(v, cnt=0):
    global ans
    cnt = cnt + 1 if cat[v] == 1 else 0
    if cnt > m:
        return
    visited.add(v)
    if len(adj[v]) == 1 and v != 1:
        ans += 1
    for i in adj[v]:
        if i not in visited:
            dfs(i, cnt)

dfs(1)
print(ans)